Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs

```diff
@@ -468,35 +479,39 @@ public class PreTypeConstraints {
       return false;
     }
 
-    public void AddConfirmation(CommonConfirmationBag check, PreType preType, IToken tok, string errorFormatString, Action onProxyAction = null) {
-      confirmations.Add(() => {
-        if (!ConfirmConstraint(check, preType, null)) {
+    public void AddConfirmation(CommonConfirmationBag check, PreType preType, IToken tok, string errorFormatString, Action onProxyAction) {
+      confirmations.Add(new ConfirmationInfo(tok,
+        () => ConfirmConstraint(check, preType, null),
+        () => string.Format(errorFormatString, preType),
+        (ResolverPass reporter) => {
           if (preType.Normalize() is PreTypeProxy && onProxyAction != null) {
             onProxyAction();
           } else {
-            PreTypeResolver.ReportError(tok, errorFormatString, preType);
+            reporter.ReportError(tok, errorFormatString, preType);
           }
-        }
-      });
+        }));
     }
 
-    public void AddConfirmation(CommonConfirmationBag check, PreType preType, Type toType, IToken tok, string errorFormatString) {
-      Contract.Requires(toType is NonProxyType);
-      var toPreType = (DPreType)PreTypeResolver.Type2PreType(toType);
-      confirmations.Add(() => {
-        if (!ConfirmConstraint(check, preType, toPreType)) {
-          PreTypeResolver.ReportError(tok, errorFormatString, preType);
-        }
-      });
-    }
-
-    public void AddConfirmation(System.Action confirm) {
-      confirmations.Add(confirm);
+    public void AddConfirmation(IToken tok, Func<bool> check, Func<string> errorMessage) {
+      confirmations.Add(new ConfirmationInfo(tok, check, errorMessage,
+        (ResolverPass reporter) => { reporter.ReportError(tok, errorMessage()); }));
     }
 
     void ConfirmTypeConstraints() {
       foreach (var confirmation in confirmations) {
-        confirmation();
+        confirmation.Confirm(PreTypeResolver);
+      }
+    }
+
+    record ConfirmationInfo(IToken Tok, Func<bool> Check, Func<string> ErrorMessage, Action<ResolverPass> OnError) {
```

Comment:
The suffix `Info` has no meaning, and there are unused public fields (Tok, Check and OnError). Please change to `record Confirmation(Action<ResolverPass> Confirm, Func<string> GetErrorMessage)`

If you want, you can introduce another method to construct this record using the `(Func<bool> Check, Func<string> ErrorMessage, Action<ResolverPass> OnError)` signature.

Reply:
They're only used internally, so they shouldn't be public fields.